packages/framework/src/Framework/Services/BuildTaskService.php

```diff
@@ -4,89 +4,122 @@
+ * @see \Hyde\Framework\Testing\Unit\BuildTaskServiceUnitTest
  */
 class BuildTaskService
 {
-    /**
-     * Information for package developers: This offers a hook for packages to add custom build tasks.
-     * Make sure to add the fully qualified class name to the array and doing so by merging the array, not overwriting it.
-     *
-     * @var array<class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>>
-     */
-    public static array $postBuildTasks = [];
+    /** @var array<string, \Hyde\Framework\Features\BuildTasks\BuildTask> */
+    protected array $buildTasks = [];
 
-    protected ?OutputStyle $output;
+    protected ?OutputStyle $output = null;
 
-    public function __construct(?OutputStyle $output = null)
+    public function __construct()
     {
-        $this->output = $output;
-    }
-
-    public function runPostBuildTasks(): void
-    {
-        foreach ($this->getPostBuildTasks() as $task) {
-            $this->run($task);
-        }
+        $this->registerTasks(Config::getArray('hyde.build_tasks', []));
 
-        $this->runIf(GenerateBuildManifest::class, config('hyde.generate_build_manifest', true));
+        $this->registerTasks($this->findTasksInAppDirectory());
     }
 
     /** @return array<class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>> */
-    public function getPostBuildTasks(): array
+    public function getTasks(): array
     {
-        return array_unique(
-            array_merge(
-                config('hyde.post_build_tasks', []),
-                static::findTasksInAppDirectory(),
-                static::$postBuildTasks
-            )
-        );
+        return array_map(function (BuildTask $task): string {
+            return $task::class;
+        }, array_values($this->buildTasks));
     }
 
-    protected static function findTasksInAppDirectory(): array
+    public function runPreBuildTasks(): void
     {
-        $tasks = [];
-
-        foreach (glob(Hyde::path('app/Actions/*BuildTask.php')) as $file) {
-            $tasks[] = str_replace(
-                [Hyde::path('app'), '.php', '/'],
-                ['App', '', '\\'],
-                (string) $file
-            );
+        foreach ($this->buildTasks as $task) {
+            if ($task instanceof RunsBeforeBuild) {
+                $task->run();
+            }
         }
+    }
 
-        return $tasks;
+    public function runPostBuildTasks(): void
+    {
+        foreach ($this->buildTasks as $task) {
+            if ($task instanceof RunsAfterBuild) {
+                $task->run();
+            }
+        }
     }
 
-    public function run(string $task): static
+    /** @param  \Hyde\Framework\Features\BuildTasks\BuildTask|class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>  $task */
+    public function registerTask(BuildTask|string $task): void
     {
-        $this->runTask(new $task($this->output));
+        $task = is_string($task) ? new $task($this->output) : $task;
```

Comment:
We can't set output here since output might not be set yet. It needs to be handled at task run time

Reply:
Additionally, with 33235ab006cb899a4316a0babcb285be739b86c9 this means that already constructed objects can't get the output-

Reply:
Resolved in ad288caea16f2c7137ed153a49e8356e3cf7498d